### PR TITLE
feat(#69): building construction

### LINKED ISSUE #69: Building Construction

### PR DESCRIPTION
**Pull request by strunk23**
Closes #69

### CODE REVIEW

File: Assets/Scripts/BuildingSystem/BuildingInfo.cs
Comment (Acu1000) [inline Assets/Scripts/BuildingSystem/BuildingInfo.cs]: don't change this; `is not` has better readability

File: Assets/Scripts/Buildings/ConstructionSite/ConstructionSite.cs
Comment (Acu1000) [inline Assets/Scripts/Buildings/ConstructionSite/ConstructionSite.cs]: Since the preview is in prototype phase, it means it'll likely be replaced in the future, possibly with a billboard GUI. Therefore, it's best if all of the ConstructionSite's resource counter visuals are handled entirely by a separate script which can be modified, reworked and replaced independently, without risking tampering with the gameplay functionality in this script.

File: Assets/Scripts/Construction/ResourceTextHandler.cs
Comment (Acu1000) [inline Assets/Scripts/Construction/ResourceTextHandler.cs]: displayText*

File: Assets/Scripts/Construction/PlayerConstructionHandler.cs
Comment (Acu1000) [inline Assets/Scripts/Construction/PlayerConstructionHandler.cs]: I'm afraid we might've had some miscommunication when discussing the task.

The current design is that adding resources to the construction is done by dropping them onto the construction site plate (resources are physical objects that the player and robots can carry around). The input system should not be used here at all

(you'll probably need to increase the grid cell size in GlobalConstants for that to work properly)

File: Assets/Scripts/Construction/PlayerConstructionHandler.cs
Comment (Acu1000) [inline Assets/Scripts/Construction/PlayerConstructionHandler.cs]: Should be serialized in BuildingInfo

**Review by Acu1000 — CHANGES_REQUESTED** [review]

File: Assets/Scripts/Buildings/ConstructionSite/ResourceTextHandler.cs
Comment (Vie0n) [inline Assets/Scripts/Buildings/ConstructionSite/ResourceTextHandler.cs]: Add empty line 
```suggestion
        private new Camera camera;

        void Start()
```

File: Assets/Scripts/Buildings/ConstructionSite/ConstructionSiteResourcePreview.cs
Comment (Vie0n) [inline Assets/Scripts/Buildings/ConstructionSite/ConstructionSiteResourcePreview.cs]: First don't nest [ternary operators](https://en.wikipedia.org/wiki/Ternary_conditional_operator)... ever.
The `transform.eulerAngles.y == 0 || transform.eulerAngles.y == 180` condition is duplicated.
I believe we can't get rid of the ifs with 90 180 and 270 ang if you use some math, mainly Sin and Cos functions from Mathf but I'm too tired (lazy) to check this so I leave it to you.

File: Assets/Scripts/Buildings/ConstructionSite/ConstructionSiteResourcePreview.cs
Comment (Vie0n) [inline Assets/Scripts/Buildings/ConstructionSite/ConstructionSiteResourcePreview.cs]: Why `1`? What does it mean? Why not `resources.Count >= 1`? Instead of using magic numbers lets define this values as some const.

File: Assets/Scripts/Buildings/ConstructionSite/ResourceTextHandler.cs
Comment (Vie0n) [inline Assets/Scripts/Buildings/ConstructionSite/ResourceTextHandler.cs]: Minor: Isn't `newAmount.ToString()` faster because it doesn't interpolate? I'm not sure if this true. @PatrykPaluch opinion?

File: Assets/Scripts/Buildings/ConstructionSite/ConstructionSite.cs
Comment (Vie0n) [inline Assets/Scripts/Buildings/ConstructionSite/ConstructionSite.cs]: Define `previewToDecrease.GetComponent<ResourcePreviewController>()` once like this
`ResourcePreviewController resourcePreviewController = previewToDecrease.GetComponent<ResourcePreviewController>()` and use it then. GetComponent is a costly operation.

**Review by Vie0n — CHANGES_REQUESTED** [review]

File: Assets/Scripts/BuildingSystem/BuildingInfo.cs
Comment (PatrykPaluch) [reply Assets/Scripts/BuildingSystem/BuildingInfo.cs]: for me `!=` is more readable than `is not` but in C# there is difference between `!=` and `is not`

- `X is not null` will check if `x` is not a `null` value
- `X != null` will use `!=` operator

in most cases you want to use `is not null`, same here

File: Assets/Scripts/Buildings/ConstructionSite/ResourceTextHandler.cs
Comment (PatrykPaluch) [reply Assets/Scripts/Buildings/ConstructionSite/ResourceTextHandler.cs]: idk, but i don't think this will have any impact on performance. `UpdateText` is called only when Resource amount changes. So in worst case scenario - one time per frame and at most ~10 times per building

File: Assets/Scripts/Construction/PlayerConstructionHandler.cs
Comment (PatrykPaluch) [reply Assets/Scripts/Construction/PlayerConstructionHandler.cs]: as @Acu1000 said, but we don't have player/robot with ability to hold items now, so just move that InputSystem logic to new script something like "DebugConstructionItemProvider" or something

Comment (strunk23) [reply Assets/Scripts/Construction/PlayerConstructionHandler.cs]: I have already reworked InputSystem so it accepts physical items being dropped onto construction site using player's gravity gun. Should I stick with previous InputSystem or will the current version suffice?

Comment (PatrykPaluch) [reply Assets/Scripts/Construction/PlayerConstructionHandler.cs]: If it is an independent script, then ok. For now we just need to mock this functionality so it's implementation doesn't matter. It will be replaced when we implement robots or player EQ

File: Assets/Scripts/Buildings/ConstructionSite/ConstructionSiteResourcePreview.cs
Comment (Acu1000) [reply Assets/Scripts/Buildings/ConstructionSite/ConstructionSiteResourcePreview.cs]: Relying on the `==` operator for floats is a bad idea in general as it can cause [floating point accuracy errors](https://en.wikipedia.org/wiki/Floating-point_arithmetic#Accuracy_problems)

Since you only represent 4 possible rotations here, I recommend using an `int` variable for it, and then multiplying it by 90 degrees when applying, like I did [here](https://github.com/grupacosmo/cosmobot/blob/0f7d00d44ab31fc103491b122125246cba6db53c/Assets/Scripts/Construction/PlayerConstructionHandler.cs#L19)

**Review by Vie0n — DISMISSED** [review]

**Comment by Vie0n** [comment]
Let @Acu1000 take a look too before you merge. He is the coordinator after all.

File: Assets/Scripts/Buildings/ConstructionSite/ConstructionSiteItemCollision.cs
Comment (Acu1000) [inline Assets/Scripts/Buildings/ConstructionSite/ConstructionSiteItemCollision.cs]: Should only consume items that are required for the construction, right now it seems to consume all of them regardless of whether or not they are needed. Check if the item is on the list and if more of it is needed

**Review by Acu1000 — CHANGES_REQUESTED** [review]

**Review by Vie0n — APPROVED** [review]

**Review by Acu1000 — APPROVED** [review]

**Pull request merged**
